flink-cdc-common/src/main/java/com/ververica/cdc/common/schema/Schema.java

```diff
@@ -0,0 +1,314 @@
+        return columns;
+    }
+
+    /** Returns all column names. It does not distinguish between different kinds of columns. */
+    public List<String> getColumnNames() {
+        return columns.stream().map(Column::getName).collect(Collectors.toList());
+    }
+
+    /**
+     * Returns all column data types. It does not distinguish between different kinds of columns.
+     */
+    public List<DataType> getColumnDataTypes() {
+        return columns.stream().map(Column::getType).collect(Collectors.toList());
+    }
+
+    /** Returns the primary keys of the table or data collection. */
+    public List<String> primaryKeys() {
+        return primaryKeys;
+    }
+
+    /** Returns the options of the table or data collection. */
+    public Map<String, String> options() {
+        return options;
+    }
+
+    /** Returns the comment of the table or data collection. */
+    public String comment() {
+        return comment;
+    }
+
+    /**
+     * Returns the {@link Column} instance for the given column name.
+     *
+     * @param columnName the name of the column
+     */
+    public Optional<Column> getColumn(String columnName) {
+        initializeNameToColumns();
+        return Optional.ofNullable(nameToColumns.get(columnName));
+    }
+
+    /**
+     * Converts all columns of this schema into a (possibly nested) row data type.
+     *
+     * @see DataTypes#ROW(DataField...)
+     */
+    public DataType toRowDataType() {
+        final DataField[] fields =
+                columns.stream().map(Schema::columnToField).toArray(DataField[]::new);
+        // the row should never be null
+        return ROW(fields).notNull();
+    }
+
+    // -----------------------------------------------------------------------------------
+    private void initializeNameToColumns() {
+        if (nameToColumns == null) {
+            // make the method thread-safe
+            synchronized (this) {
+                // we need to check nullability again here
+                if (nameToColumns == null) {
+                    nameToColumns = new HashMap<>();
+                    for (Column col : columns) {
+                        nameToColumns.put(col.getName(), col);
+                    }
+                }
+            }
+        }
+    }
+
+    private static DataField columnToField(Column column) {
+        return DataTypes.FIELD(column.getName(), column.getType());
+    }
+
+    /** Builder for configuring and creating instances of {@link Schema}. */
+    public static Schema.Builder newBuilder() {
+        return new Builder();
+    }
+
+    // -----------------------------------------------------------------------------------
+
+    /**
+     * A builder for constructing an immutable but still unresolved {@link
+     * org.apache.flink.table.api.Schema}.
+     */
+    @PublicEvolving
+    public static final class Builder {
+
+        private final List<Column> columns;
+        private final Set<String> columnNames;
```

Comment:
How about use `private final Map<String, Column>` here?
Then we can change the `Schema`'s constructor as  `Schema(Map<String, Column> columns, ...)`

Reply:
It's okay to introduce the map in Builder, but keeping `Map<String, Column>` in `Schema` is a little wired it's implementation need instead of API need. Btw, `Column` contains its `name` field which may be error-prone (maybe different with the key in the map if we only maintain all columns of `schema`)

Reply:
> It's okay to introduce the map in Builder, but keeping `Map<String, Column>` in `Schema` is a little wired it's implementation need instead of API need. Btw, `Column` contains its `name` field which may be error-prone (maybe different with the key in the map if we only maintain all columns of `schema`)

+1 , Build `Map<String, Column>` inner will be better.